### PR TITLE
Explicitly disable groupby on unsupported key types.

### LINKED ISSUE #8905: [BUG] `groupby` on `STRUCT` / `LIST` input passes silently, returns incorrect grouping.
When a `groupby` is constructed on a `STRUCT` or `LIST` input, it does not currently fail, as one might expect of an unsupported type.

The `groupby.aggregate()` seems to return the input groupby column unchanged, implying that each row is its own group. I think this also contributes to issues like #8887, where the `grouped_rolling_window()` returns unexpected output.

It might be worth considering a `CUDF_FAIL()` for unsupported input types, at least until support is added.

The repro case for `STRUCT` input follows:

```c++
TEST_F(UtilitiesTest, groupby_struct_bork)
{
  using namespace cudf;
  using ints = fixed_width_column_wrapper<int32_t>;

  auto child       = ints{1, 2, 3, 1, 2, 2, 1, 3, 3, 2};
  auto struct_keys = structs_column_wrapper{child};
  ints vals{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};

  auto expect_child       = ints{1, 2, 3};
  auto expect_struct_keys = structs_column_wrapper{expect_child};
  ints expect_vals{9, 19, 17};

  std::cout << "Struct input to groupby (keys)!" << std::endl;
  print(struct_keys);

  std::cout << "Expected output (grouped) struct keys!" << std::endl;
  print(expect_struct_keys);

  auto requests = std::vector<groupby::aggregation_request>{};
  requests.emplace_back(groupby::aggregation_request{});
  auto& agg_request  = requests.front();
  agg_request.values = vals;
  agg_request.aggregations.push_back(cudf::make_sum_aggregation());
  auto gby = groupby::groupby{table_view{{struct_keys}}, null_policy::EXCLUDE, sorted::NO, {}, {}};
  auto result = gby.aggregate(requests);

  std::cout << "Actual output (grouped) struct keys: " << std::endl;
  print(result.first->view().column(0));
}
```

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #8905.

Attempting groupby aggregations with `LIST` keys leads to silent
failures and bad results.
For instance, attempting hash-based `groupby` aggregations with `LIST`
keys only fails on DEBUG builds, thus:
```
/home/myth/dev/cudf/2/cpp/include/cudf/table/row_operators.cuh:447: unsigned int cudf:
:element_hasher_with_seed<hash_function, has_nulls>::operator()(cudf::column_device_view, signed in
t) const [with T = cudf::list_view; void *<anonymous> = (void *)nullptr; hash_function = default_ha
sh; __nv_bool has_nulls = false]: block: [0,0,0], thread: [0,0,0] Assertion `false && "Unsupported
type in hash."` failed.
```
In RELEASE builds, a copy of the input `LIST` column is returned, causing
each output row to be interpreted as its own group.

This commit adds an explicit failure for unsupported groupby key types,
i.e. those that don't support equality comparisons (like `LIST`).

### CODE REVIEW

File: cpp/src/groupby/groupby.cu [hunk 1]
```diff
@@ -55,6 +55,9 @@ groupby::groupby(table_view const& keys,
     _column_order{column_order},
     _null_precedence{null_precedence}
 {
+  auto is_list  = [](auto const& col) { return col.type().id() == cudf::type_id::LIST; };
+  auto has_list = std::any_of(keys.begin(), keys.end(), is_list);
+  CUDF_EXPECTS(not has_list, "Groupby does not support LIST columns as keys.");
```
Comment (jrhemstad) [inline cpp/src/groupby/groupby.cu:60]: This check should be put in a more appropriate place. Perhaps the constructor of the `row_equality_comparator` which should verify that all of the columns passed in are type that is equality comparable.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:60]: Oh, yes. Far more elegant.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:60]: Hmm. A minor wrinkle. Here's the signature for `row_equality_comparator`'s constructor:
```c++
  row_equality_comparator(table_device_view lhs, table_device_view rhs, bool nulls_are_equal = true)
    : lhs{lhs}, rhs{rhs}, nulls_are_equal{nulls_are_equal}
  {...}
```
This works on `table_device_view`, which does not allow access/iteration on the columns from `__host__`. :/

File: cpp/include/cudf/table/row_operators.cuh [hunk 2]
```diff
@@ -222,13 +223,75 @@ class element_equality_comparator {
...
+/**
+ * @brief Checks if corresponding columns in `lhs` and `rhs` tables may be compared for equality
+ *
+ * @param lhs Left table, whose columns are to be compared for equality.
+ * @param rhs Right table, whose columns are to be compared for equality.
+ * @return true If *all* columns in `lhs` permit equality comparisons with corresponding columns in
+ * `rhs`.
+ * @return false Otherwise.
+ */
+bool all_columns_are_equality_comparable(table_device_view lhs, table_device_view rhs)
```
Comment (mythrocks) [inline cpp/include/cudf/table/row_operators.cuh:274]: TODO: This should probably be moved out of line, into a `.cu` file.

File: cpp/include/cudf/table/row_operators.cuh [hunk 3]
```diff
@@ -222,13 +223,75 @@ class element_equality_comparator {
...
+
+/**
+ * @brief Checks if two columns have types that permit equality comparisons
+ *
+ * @param lhs Left column for equality comparisons
+ * @param rhs Right column for equality comparisons
+ * @return true If `lhs` and `rhs` columns have types whose values may be compared for equality
+ * @return false Otherwise.
+ */
+bool __device__ columns_are_equality_comparable(column_device_view lhs, column_device_view rhs)
```
Comment (jrhemstad) [inline cpp/include/cudf/table/row_operators.cuh:260]: I don't think this should be a device function.

File: cpp/include/cudf/table/row_operators.cuh [hunk 4]
```diff
@@ -222,13 +223,75 @@ class element_equality_comparator {
...
+ * @return false Otherwise.
+ */
+bool all_columns_are_equality_comparable(table_device_view lhs, table_device_view rhs)
+{
+  auto columns_begin = thrust::make_counting_iterator<size_type>(0);
+  auto columns_end   = columns_begin + lhs.num_columns();
+  auto eq_comparable = [lhs, rhs] __device__(auto col_idx) {
+    return columns_are_equality_comparable(lhs.column(col_idx), rhs.column(col_idx));
+  };
+  return thrust::all_of(thrust::seq, columns_begin, columns_end, eq_comparable);
```
Comment (jrhemstad) [inline cpp/include/cudf/table/row_operators.cuh:281]: So the lambda is a device lambda, but this is running in host code. Furthermore, we shouldn't be launching a kernel to check this. That's expensive.

File: cpp/include/cudf/table/row_operators.cuh (hunk 3 above)
Comment (mythrocks) [reply cpp/include/cudf/table/row_operators.cuh:260]: But isn't `table_device_view::column(size_type)` a `__device__` function?

File: cpp/include/cudf/table/row_operators.cuh (hunk 4 above)
Comment (mythrocks) [reply cpp/include/cudf/table/row_operators.cuh:281]: `lhs.column(col_idx)` is a `__device__` function. 

The only recourse I can think of is to change `row_equality_comparator` to operate on `table_view`. That looks deeply intrusive. There are a lot of call sites that seem to wrap tightly around the `table_device_view` bit.

Comment (jrhemstad) [reply cpp/include/cudf/table/row_operators.cuh:281]: Agreed. Lets defer the `row_equality_comparator` changes, but keep the `equality_comparable` check and just put it at a higher level in groupby for now.

File: cpp/src/groupby/groupby.cu [hunk 5]
```diff
@@ -43,6 +43,19 @@
...
+ * @brief Helper functor to check if a specified type supports equality comparisons.
+ */
+struct equality_comparable_functor {
+  template <typename T>
+  bool operator()() const
+  {
+    return cudf::is_equality_comparable<T, T>();
+  }
+};
+}  // namespace
```
Comment (jrhemstad) [inline cpp/src/groupby/groupby.cu:57]: Make a "runtime" trait for `data_type` that wraps this with the `type_dispacher` in `traits.hpp`.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/groupby/groupby.cu (hunk 5 above)
Comment (jrhemstad) [reply cpp/src/groupby/groupby.cu:57]: See https://github.com/rapidsai/cudf/blob/0cf2c5ff32daca70e58c2acfc78ca28fc333ba18/cpp/include/cudf/utilities/traits.hpp#L158-L180 for example

Comment (ttnghia) [reply cpp/src/groupby/groupby.cu:57]: Ah yes, this check is very useful for general usage too.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:57]: Hey, @jrhemstad. Sorry, I haven't grokked your advice here. Please ELI5?

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:57]: Oh, I see... Make an `is_equality_comparable(column_view, column_view)` available in `traits.hpp`, and use that here?

Comment (jrhemstad) [reply cpp/src/groupby/groupby.cu:57]: Not quite. `is_equality_comparable(data_type)` is what I meant.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:57]: I should really look into why GitHub notifications arrive so late.)

I tried writing a binary version of the same (with differing `lhs` and `rhs` type). That would require double dispatch, even for a single type.
(That was a long walk down a winding road to a cafe that was closed. :p) I'll have the unary version up shortly.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu]: This is sorted out. I've left the binary version of `is_equality_comparable()` out of this PR, since it's not necessary for this change.

**Comment by mythrocks** [comment]
Taking @cwharris off the hook. Thanks for taking up the review in his stead, @ttnghia.

**Comment by mythrocks** [comment]
The failures seem transient:
```
02:25:26   fatal: unable to access 'https://github.com/dask/distributed.git/': The requested URL returned error: 429
02:25:26 WARNING: Discarding git+https://github.com/dask/distributed.git@main. Command errored out with exit status 128: git clone -q https://github.com/dask/distributed.git /tmp/pip-req-build-m_6j6ejd Check the logs for full command output.
```

**Comment by mythrocks** [comment]
Rerun tests

File: cpp/src/groupby/groupby.cu [hunk 6]
```diff
@@ -55,6 +56,13 @@ groupby::groupby(table_view const& keys,
     _column_order{column_order},
     _null_precedence{null_precedence}
 {
+  auto is_supported_key_type = [](auto col) {
+    return col.type().id() == type_id::STRUCT     // Struct does not support ==.
+           or col.type().id() == type_id::STRING  // String does not support == on __host__.
```
Comment (jrhemstad) [inline cpp/src/groupby/groupby.cu:61]: Special case logic like this is precisely what I was hoping to avoid with `is_equality_comparable()`. Furthermore, this doesn't seem right. What if the STRUCT column contains a column where `is_equality_comparable()` fails? This check should happen after the flattening.

How does `is_equality_comparable(STRING)` fail?

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu:61]: > this doesn't seem right. What if the STRUCT column contains a column where is_equality_comparable() fails?

That check (for `STRUCT` containing `LIST`) exists already, as part of `flatten_nested_columns()`.
Hmm. After the last round of changes to the `flatten()`/`unflatten()` code path, I wonder if we should simply rely on `flatten()` flagging unsupported types. I'll check this in a `Release` build.

> How does `is_equality_comparable(STRING)` fail?

That's an unfortunate consequence: `string_view::operator ==` is `__device__`, while `row_equality_operator` is checking `is_equality_comparable()` from `__host__`. This causes `row_equality_operator` to bork on `STRING`.

Comment (mythrocks) [reply cpp/src/groupby/groupby.cu]: This has been resolved, by moving the checks to after `flatten_nested_columns()`.

File: cpp/src/groupby/common/utils.cpp [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+#include <cudf/utilities/traits.hpp>
+
+namespace cudf::groupby::detail {
+
+void assert_keys_equality_comparable(cudf::table_view const& keys)
+{
+  auto is_supported_key_type = [](auto col) { return cudf::is_equality_comparable(col.type()); };
+  CUDF_EXPECTS(std::all_of(keys.begin(), keys.end(), is_supported_key_type),
+               "Unsupported groupby key type does not support equality comparison");
+}
```
Comment (jrhemstad) [inline cpp/src/groupby/common/utils.cpp:29]: tbh, I'm going to take inspiration from @shwina's latest Better Code talk about "don't write functions" and say that it's fine to just repeat this logic inline.  We know that ultimately this check will move into `row_equality_comparator`'s ctor.

Comment (mythrocks) [reply cpp/src/groupby/common/utils.cpp:29]: Drat. My money was on you voting the other way. :]

I'll update.

Comment (mythrocks) [reply cpp/src/groupby/common/utils.cpp]: I've moved this code inline, now.

File: cpp/src/groupby/common/utils.hpp [hunk 8]
```diff
@@ -39,6 +39,14 @@ inline std::vector<aggregation_result> extract_results(host_span<RequestType con
   return results;
 }
 
+/**
+ * @brief Fails an assert if any of the specified `keys` columns do not support equality
+ * comparisons.
+ *
+ * @param keys Grouping keys to be checked for equality comparisons.
+ */
+void assert_keys_equality_comparable(cudf::table_view const& keys);
```
Comment (jrhemstad) [inline cpp/src/groupby/common/utils.hpp:48]: This looks leftover?

**Review by jrhemstad — APPROVED** [review]

**Comment by mythrocks** [comment]
Thanks for the reviews, all. I'll merge this in now.

**Comment by mythrocks** [comment]
@gpucibot merge
